### ISSUE TITLE
nvptx broken > llvm 3.7

### ISSUE THREAD

**Issue by svenevs** [ispc/ispc#1251]
Hello, I'm working off of tag `v1.9.1` and am using `llvm 3.9.0`.  I noticed that the special treatment for updated functions in `llvm` was overlooked in the NVPTX section of `opt.cpp`.  I would ordinarily submit a PR with this, but honestly I don't know if it works.

The diff (to reiterate, **from tag v1.9.1**):
```
--- a/opt.cpp
+++ b/opt.cpp
@@ -805,8 +805,13 @@ Optimize(llvm::Module *module, int optLevel) {
           optPM.add(CreatePromoteLocalToPrivatePass());
           optPM.add(llvm::createGlobalDCEPass());
 
+#if ISPC_LLVM_VERSION <= ISPC_LLVM_3_7
           optPM.add(llvm::createTypeBasedAliasAnalysisPass());
           optPM.add(llvm::createBasicAliasAnalysisPass());
+#else
+          optPM.add(llvm::createTypeBasedAAWrapperPass());
+          optPM.add(llvm::createBasicAAWrapperPass());
+#endif
           optPM.add(llvm::createCFGSimplificationPass());
           // Here clang has an experimental pass SROAPass instead of
           // ScalarReplAggregatesPass. We should add it in the future.
@@ -817,8 +822,14 @@ Optimize(llvm::Module *module, int optLevel) {
 #endif
           optPM.add(llvm::createEarlyCSEPass());
           optPM.add(llvm::createLowerExpectIntrinsicPass());
+
+#if ISPC_LLVM_VERSION <= ISPC_LLVM_3_7
           optPM.add(llvm::createTypeBasedAliasAnalysisPass());
           optPM.add(llvm::createBasicAliasAnalysisPass());
+#else
+          optPM.add(llvm::createTypeBasedAAWrapperPass());
+          optPM.add(llvm::createBasicAAWrapperPass());
+#endif
 
           // Early optimizations to try to reduce the total amount of code to
           // work with if we can
@@ -846,7 +857,17 @@ Optimize(llvm::Module *module, int optLevel) {
           optPM.add(llvm::createInstructionCombiningPass());
           optPM.add(llvm::createCFGSimplificationPass());
           optPM.add(llvm::createPruneEHPass());
+
+#if ISPC_LLVM_VERSION >= ISPC_LLVM_3_9 // 3.9+
+          optPM.add(llvm::createPostOrderFunctionAttrsLegacyPass());
+          optPM.add(llvm::createReversePostOrderFunctionAttrsPass());
+#elif ISPC_LLVM_VERSION == ISPC_LLVM_3_8 // 3.8
+          optPM.add(llvm::createPostOrderFunctionAttrsPass());
+          optPM.add(llvm::createReversePostOrderFunctionAttrsPass());
+#else // 3.7 and earlier
           optPM.add(llvm::createFunctionAttrsPass());
+#endif
+
           optPM.add(llvm::createFunctionInliningPass());
           optPM.add(llvm::createConstantPropagationPass());
           optPM.add(llvm::createDeadInstEliminationPass());
```
I basically just copied what you all did elsewhere in the file when these functions were being called.

**Edit**:

The above patch appears to be working, but I don't really know how to test this.  I was being bad about testing this, the `Cannot select: intrinsic %llvm.x86.avx2.gather.d.pd.256` occurs reliably when `ARM_ENABLED=1`.

So long story short, if it makes your life easier, I'll submit a PR with the above for nvptx...I don't work with arm so I'm going to happily ignore it ;)

~After applying that, it will compile an `ispc` executable, but it is broken complaining about `LLVM ERROR: Cannot select: intrinsic %llvm.x86.avx2.gather.d.pd.256` (e.g. for the `examples/gmres`).  Noting that this all came from a `make NVPTX_ENABLED=1`, if I compile it without this then the executable can be used to compile the example gmres program just fine.~

~This tool / working with `llvm` is very new for me, apologies if the error has a more pointed / obvious meaning...but thanks for any pointers!~

**Comment by egaburov** [comment]
I attempted to test the patch, but I'm getting the following error when trying to building examples, even when building ispc with `NVPTX_ENABLED=0`:

```
examples/gmres$ make
/bin/mkdir -p objs/
ispc -O2 --arch=x86-64 --target=sse2-i32x4,sse4-i32x8,avx1-i32x16,avx2-i32x16,avx512knl-i32x16,avx512skx-i32x16 matrix.ispc -o objs/matrix_ispc.o -h objs/matrix_ispc.h
*** Error in `ispc': munmap_chunk(): invalid pointer: 0x0000000005bb9b40 ***
main.cpp(235): FATAL ERROR: Unhandled signal sent to process; terminating.
***
*** Please file a bug report at https://github.com/ispc/ispc/issues
*** (Including as much information as you can about how to reproduce this error).
*** You have apparently encountered a bug in the compiler that we'd like to fix!
***
make: *** [objs/matrix_ispc.h] Aborted (core dumped)
make: *** Deleting file `objs/matrix_ispc.h'
```
Similar error occurs when I build mandelbrot.

Config: Ubuntu 14.04,  llvm-3.9 (built with alloy.py -b --version=3.9), ispc master branch. 

@dbabokin any clue what might be going on? Thanks

**Comment by svenevs** [comment]
Closing as stale.
